kge/dataset.py

```diff
@@ -209,44 +213,45 @@ def _load_map(
-            pickle_suffix = Dataset._to_valid_filename(
-                f"-{as_list}-{delimiter}-{ignore_duplicates}.pckl"
-            )
-            pickle_filename = filename + pickle_suffix
-            result = Dataset._pickle_load_if_uptodate(None, pickle_filename, filename)
-            if result is not None:
-                return result
-
-        n = 0
-        dictionary = {}
-        warned_overrides = False
-        duplicates = 0
-        with open(filename, "r") as file:
-            for line in file:
-                key, value = line.split(delimiter, maxsplit=1)
-                value = value.rstrip("\n")
-                if as_list:
-                    key = int(key)
-                    n = max(n, key + 1)
-                if key in dictionary:
-                    duplicates += 1
-                    if not ignore_duplicates:
-                        raise KeyError(f"{filename} contains duplicated keys")
-                else:
-                    dictionary[key] = value
-        if as_list:
-            array = [None] * n
-            for index, value in dictionary.items():
-                array[index] = value
-            result = (array, duplicates)
-        else:
-            result = (dictionary, duplicates)
+        with Dataset.lock:
+            if use_pickle:
+                # check if there is a pickled, up-to-date version of the file
+                pickle_suffix = Dataset._to_valid_filename(
+                    f"-{as_list}-{delimiter}-{ignore_duplicates}.pckl"
+                )
+                pickle_filename = filename + pickle_suffix
+                result = Dataset._pickle_load_if_uptodate(None, pickle_filename, filename)
+                if result is not None:
+                    return result
+
+            n = 0
+            dictionary = {}
+            warned_overrides = False
+            duplicates = 0
+            with open(filename, "r") as file:
+                for line in file:
+                    key, value = line.split(delimiter, maxsplit=1)
+                    value = value.rstrip("\n")
+                    if as_list:
+                        key = int(key)
+                        n = max(n, key + 1)
+                    if key in dictionary:
+                        duplicates += 1
+                        if not ignore_duplicates:
+                            raise KeyError(f"{filename} contains duplicated keys")
+                    else:
+                        dictionary[key] = value
+            if as_list:
+                array = [None] * n
+                for index, value in dictionary.items():
+                    array[index] = value
+                result = (array, duplicates)
+            else:
+                result = (dictionary, duplicates)
 
-        if use_pickle:
-            Dataset._pickle_dump_atomic(result, pickle_filename)
-        return result
+            if use_pickle:
+                Dataset._pickle_dump_atomic(result, pickle_filename)
```

Comment:
and _pickle_dump_atomic? Then multiple processes can still work in parallel, but loading/writing of pickles files is sequential.

Reply:
and we could do the writing in the background as well, since it takes quite long for large indexes and there is no need to wait for it.

Reply:
You mean only protect the "move file" part? Sounds good to me.

Reply:
yes, protect the move file part and call _pickle_dump_atomic in a separate thread, so that we can compute the next index already while we write to disk.
The writing process does not need to wait for the writing to finish, since it will never read it.

Reply:
That's too much optimzation for a bugfix PR ;). For now, I suggest to only protect move (but not the dumping) but do not start adding threading.